serialization/date/unmarshal_utils.go

```diff
@@ -0,0 +1,405 @@
+package date
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"time"
+)
+
+const (
+	negInt64       = int64(-1) << 32
+	zeroDate       = "-5877641-06-23"
+	zeroMS   int64 = -185542587187200000
+)
+
+var errWrongDataLen = fmt.Errorf("failed to unmarshal date: the length of the data should be 0 or 4")
+
+func errNilReference(v interface{}) error {
+	return fmt.Errorf("failed to unmarshal date: can not unmarshal into nil reference (%T)(%[1]v))", v)
+}
+
+func DecInt32(p []byte, v *int32) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		*v = 0
+	case 4:
+		*v = decInt32(p)
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecInt32R(p []byte, v **int32) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = nil
+		} else {
+			*v = new(int32)
+		}
+	case 4:
+		val := decInt32(p)
+		*v = &val
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecInt64(p []byte, v *int64) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		*v = zeroMS
+	case 4:
+		*v = decMilliseconds(p)
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecInt64R(p []byte, v **int64) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = nil
+		} else {
+			val := zeroMS
+			*v = &val
+		}
+	case 4:
+		val := decMilliseconds(p)
+		*v = &val
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecUint32(p []byte, v *uint32) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		*v = 0
+	case 4:
+		*v = decUint32(p)
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecUint32R(p []byte, v **uint32) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = nil
+		} else {
+			*v = new(uint32)
+		}
+	case 4:
+		val := decUint32(p)
+		*v = &val
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecString(p []byte, v *string) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = ""
+		} else {
+			*v = zeroDate
+		}
+	case 4:
+		*v = decString(p)
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecStringR(p []byte, v **string) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = nil
+		} else {
+			val := zeroDate
+			*v = &val
+		}
+	case 4:
+		val := decString(p)
+		*v = &val
+	default:
+		return errWrongDataLen
+	}
+	return nil
+}
+
+func DecTime(p []byte, v *time.Time) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	switch len(p) {
+	case 0:
+		if p == nil {
+			*v = time.Time{}
+		} else {
+			*v = minTime.UTC()
```

Comment:
it is already in UTC:
```suggestion
			*v = minTime
```

Reply:
This was done to avoid unnecessary dependence, because in the `time.Time` struct the `loc` field stores reference on `time.Location`.

Reply:
I remove are UTC() method appliances.